compiler/src/iree/compiler/Codegen/Common/GPU/GPUHeuristics.cpp

```diff
@@ -17,10 +18,67 @@ using llvm::APIntOps::GreatestCommonDivisor;
 
 namespace mlir::iree_compiler {
 
+int64_t sharedMemoryUsed(const GPUMMASchedule &schedule, int64_t lhsBitwidth,
+                         int64_t rhsBitwidth) {
+  int64_t tileM = schedule.mSize * schedule.mTileCount * schedule.mWarpCount;
+  int64_t tileN = schedule.nSize * schedule.nTileCount * schedule.nWarpCount;
+  int64_t tileK = schedule.kSize * schedule.kTileCount;
+  return tileM * tileK * lhsBitwidth + tileN * tileK * rhsBitwidth;
+}
+
+std::optional<GPUMMASchedule>
+fitScheduleInSharedMemory(ArrayRef<GPUMatmulShapeType> intrinsics,
+                          GPUMMASchedule schedule, int64_t sharedMemLimit) {
+  int64_t lhsBitwidth =
+      intrinsics[schedule.index].aType.getIntOrFloatBitWidth() / 8;
+  int64_t rhsBitwidth =
+      intrinsics[schedule.index].bType.getIntOrFloatBitWidth() / 8;
+  while (sharedMemoryUsed(schedule, lhsBitwidth, rhsBitwidth) >
+         sharedMemLimit) {
+
+    LLVM_DEBUG({
+      llvm::errs() << "Shrinking schedule\n";
+      llvm::errs() << "mSize: " << schedule.mSize << "\n";
+      llvm::errs() << "nSize: " << schedule.nSize << "\n";
+      llvm::errs() << "kSize: " << schedule.kSize << "\n";
+      llvm::errs() << "mTileCount: " << schedule.mTileCount << "\n";
+      llvm::errs() << "nTileCount: " << schedule.nTileCount << "\n";
+      llvm::errs() << "kTileCount: " << schedule.kTileCount << "\n";
+      llvm::errs() << "mWarpCount: " << schedule.mWarpCount << "\n";
+      llvm::errs() << "nWarpCount: " << schedule.nWarpCount << "\n";
+    });
+    // Attempt to shrink the schedule along one of the dimensions.
+    if (schedule.mTileCount > 1) {
+      schedule.mTileCount--;
+      continue;
+    }
+    if (schedule.nTileCount > 1) {
+      schedule.nTileCount--;
+      continue;
+    }
+    if (schedule.kTileCount > 1) {
+      schedule.kTileCount--;
```

Comment:
This won't work right now. If my dimension size is `1024` and I pick `tileCount = 4` and `warpCount = 4` for an intrinsic size of `16`, then the tile size we get will be `256`, which nicely divides `1024`. If we decrease the tile size by halving a tile count, we'll get a resulting tile size of `128`, however if we do it by decrementing `tileCount`, we'll get `3 * 4 * 16 = 192` which does not properly divide the problem size of `1024`.

Reply:
Right. I reverted it for now, but an easy fix would also be to simply check if its a valid schedule or not. Going to leave it as a TODO.